File: web/src/services/api/authSession.ts
import { api } from '@/services/api/api';
import { storageAuthToken } from '@/storage/storageAuthToken';
import { storageRefreshToken } from '@/storage/storageRefreshToken';

export interface AuthTokensResponse {
  token: string;
  refreshToken: string;
  expiresIn: number;
}

let refreshPromise: Promise<string | null> | null = null;

export function applyAccessToken(accessToken: string, expiresInSeconds: number): void {
  storageAuthToken.setAuthToken(accessToken);
  storageRefreshToken.setAccessExpiresAt(expiresInSeconds);
  api.defaults.headers.common.Authorization = `Bearer ${accessToken}`;
}

export function persistAuthTokens(data: AuthTokensResponse): void {
  applyAccessToken(data.token, data.expiresIn);
  storageRefreshToken.setRefreshToken(data.refreshToken);
}

export function clearAuthSession(): void {
  storageAuthToken.removeAuthToken();
  storageRefreshToken.removeRefreshToken();
  storageRefreshToken.removeAccessExpiresAt();
  delete api.defaults.headers.common.Authorization;
}

export async function refreshAccessToken(): Promise<string | null> {
  if (refreshPromise) {
    return refreshPromise;
  }

  refreshPromise = (async () => {
    const refreshToken = storageRefreshToken.getRefreshToken();
    if (!refreshToken) {
      return null;
    }

    try {
      const response = await api.post<AuthTokensResponse>('/refresh', {
        refreshToken,
      });
      persistAuthTokens(response.data);
      return response.data.token;
    } catch {
      clearAuthSession();
      return null;
    } finally {
      refreshPromise = null;
    }
  })();

  return refreshPromise;
}

export async function getValidAccessToken(): Promise<string | null> {
  const stored = storageAuthToken.getAuthToken();
  if (stored && !storageRefreshToken.isAccessTokenExpiringSoon()) {
    api.defaults.headers.common.Authorization = `Bearer ${stored}`;
    return stored;
  }
  return refreshAccessToken();
}

export async function logoutRemote(): Promise<void> {
  const refreshToken = storageRefreshToken.getRefreshToken();
  if (refreshToken) {
    try {
      await api.post('/logout', { refreshToken });
    } catch {
      // ignora falha remota
    }
  }
  clearAuthSession();
}

File: mobile/lib/services/auth/auth_session.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:planejamento_urbano/config/api_constants.dart';
import 'package:planejamento_urbano/storage/storage_refresh_token.dart';
import 'package:planejamento_urbano/storage/storage_token.dart';

/// Renovação de tokens e sessão (login, refresh, logout).
class AuthSession {
  AuthSession._();

  static Future<String?>? _refreshInFlight;

  static Future<void> persistLoginResponse(Map<String, dynamic> data) async {
    final access = data['token'] as String?;
    final refresh = data['refreshToken'] as String?;
    final expiresIn = (data['expiresIn'] as num?)?.toInt() ?? 3600;

    if (access == null || refresh == null) {
      throw Exception('Resposta de autenticação incompleta.');
    }

    await StorageToken.storageAuthTokenSave(access);
    await StorageRefreshToken.save(refresh, expiresIn);
  }

  static Future<void> clearSession() async {
    await StorageToken.storageAuthTokenRemove();
    await StorageRefreshToken.remove();
  }

  static Future<String?> getValidAccessToken() async {
    final access = await StorageToken.storageAuthTokenGet();
    if (access != null && !await StorageRefreshToken.isAccessTokenExpiringSoon()) {
      return access;
    }
    return refreshAccessToken();
  }

  static Future<String?> refreshAccessToken() async {
    if (_refreshInFlight != null) {
      return _refreshInFlight;
    }

    _refreshInFlight = _refreshInternal();
    try {
      return await _refreshInFlight;
    } finally {
      _refreshInFlight = null;
    }
  }

  static Future<String?> _refreshInternal() async {
    final refresh = await StorageRefreshToken.get();
    if (refresh == null) return null;

    final response = await http.post(
      Uri.parse('$baseUrl/refresh'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'refreshToken': refresh}),
    );

    if (response.statusCode != 200) {
      await clearSession();
      return null;
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    await persistLoginResponse(data);
    return data['token'] as String?;
  }

  static Future<void> logoutRemote() async {
    final refresh = await StorageRefreshToken.get();
    if (refresh != null) {
      try {
        await http.post(
          Uri.parse('$baseUrl/logout'),
          headers: {'Content-Type': 'application/json'},
          body: jsonEncode({'refreshToken': refresh}),
        );
      } catch (_) {
        // ignora falha remota
      }
    }
    await clearSession();
  }
}
